### PR TITLE
Add unit tests to validate error states are detected

### PR DESCRIPTION
**Pull request by sawsa307**
Add TestParseEndpointBatchErr to verify invalid endpoint batch error state. Add TestValidateEndpointFields to verify invalid endpoint information error state.

### CODE REVIEW

**Comment by sawsa307** [comment]
/hold
Rebase after #1963 and #2039 get merged

File: pkg/neg/syncers/transaction_test.go [hunk 1]
```diff
@@ -1869,6 +1869,68 @@ func TestEnableDegradedMode(t *testing.T) {
...
+	for _, tc := range testCases {
+		t.Run(tc.desc, func(t *testing.T) {
+			mockGCE := fakeGCE.Compute().(*cloud.MockGCE)
+			mockGCE.MockNetworkEndpointGroups.AttachNetworkEndpointsHook = func(ctx context2.Context, key *meta.Key, arg0 *compute.NetworkEndpointGroupsAttachEndpointsRequest, neg *cloud.MockNetworkEndpointGroups) error {
+				return tc.returnError
+			}
+			_, transactionSyncer := newTestTransactionSyncer(fakeCloud, negtypes.VmIpPortEndpointType, false)
+
+			err := transactionSyncer.cloud.AttachNetworkEndpoints(transactionSyncer.NegSyncerKey.NegName, zone, networkEndpoints, transactionSyncer.NegSyncerKey.GetAPIVersion())
+			err = checkEndpointBatchErr(err, tc.endpointOperation)
```
Comment (swetharepakula) [inline pkg/neg/syncers/transaction_test.go:1926]: Can we not directly pass in the return error here?

Why do we need the transaction syncer and the mock attachnetwork endpoint hook?

File: pkg/neg/syncers/utils_test.go [hunk 2]
```diff
@@ -731,6 +731,371 @@ func TestIpsForPod(t *testing.T) {
...
+				negtypes.TestZone2: negtypes.NewNetworkEndpointSet(
+					networkEndpointFromEncodedEndpoint("10.100.3.1||instance3||80")),
+			},
+			expectMap: negtypes.EndpointPodMap{
+				networkEndpointFromEncodedEndpoint("10.100.3.1||instance3||80"): types.NamespacedName{Namespace: testServiceNamespace, Name: "pod4"},
+			},
+			expectErr: nil,
+		},
+		{
+			desc:     "include one endpoint that does correspond to an existing node",
```
Comment (swetharepakula) [inline pkg/neg/syncers/utils_test.go:997]: nit: does not correspond

File: pkg/neg/syncers/utils_test.go [hunk 3]
```diff
@@ -731,6 +731,371 @@ func TestIpsForPod(t *testing.T) {
...
+							Name:     &emptyNamedPort,
+							Port:     &port80,
+							Protocol: &protocolTCP,
+						},
+					},
+				},
+			},
+			expectSets: nil,
+			expectMap:  nil,
+			expectErr:  negtypes.ErrEPZoneMissing,
```
Comment (swetharepakula) [inline pkg/neg/syncers/utils_test.go:1038]: shouldn't this be ErrEPNodeNotFound ?

File: pkg/neg/syncers/utils_test.go (hunk 2 above)
Comment (sawsa307) [reply pkg/neg/syncers/utils_test.go:997]: Updated. Thanks!

File: pkg/neg/syncers/utils_test.go (hunk 3 above)
Comment (sawsa307) [reply pkg/neg/syncers/utils_test.go:1038]: I want to reserve `ErrEPNodeNotFound` for the error when we cannot find a node from nodeLister. Here we are checking zoneGetter to see if we have an entry for this node.

File: pkg/neg/syncers/transaction.go [hunk 4]
```diff
@@ -523,8 +523,9 @@ func (s *transactionSyncer) operationInternal(operation transactionOp, zone stri
 		// we would set error state and retry. For successful calls, we won't update
 		// error state, so its value won't be overwritten within API call go routines.
 		if syncErr.IsErrorState {
-			s.logger.V(3).Info("Updating error state", "error state", syncErr.Reason)
+			s.logger.Info("Detected unexpected error when checking endpoint update response", "operation", operation, "error", err)
 			s.syncLock.Lock()
+			s.logger.V(3).Info("Updating error state", "error state", syncErr.Reason)
```
Comment (swetharepakula) [inline pkg/neg/syncers/transaction.go:528]: nit: should you remove the `V(3)` here too?

**Review by swetharepakula — COMMENTED** [review]
Some small nits, otherwise LGTM

**Review by swetharepakula — COMMENTED** [review]
Some small nits, otherwise LGTM

File: pkg/neg/syncers/utils.go [hunk 5]
```diff
@@ -436,7 +436,7 @@ func toZoneNetworkEndpointMapDegradedMode(eds []negtypes.EndpointsData, zoneGett
 			}
 			zone, getZoneErr := zoneGetter.GetZoneForNode(nodeName)
 			if getZoneErr != nil {
-				klog.Errorf("For endpoint %q in pod %q, its corresponding node %q does not have valid zone information: %w, skipping", endpointAddress.Addresses, pod.ObjectMeta.Name, nodeName, getZoneErr)
+				klog.Errorf("For endpoint %q in pod %q, its corresponding node %q does not have valid zone information: %v, skipping", endpointAddress.Addresses, pod.ObjectMeta.Name, nodeName, getZoneErr)
```
[the remark this answers is not in the document]
Comment (swetharepakula) [reply pkg/neg/syncers/utils.go:439]: Ahh, okay missed that it was klog.Errorf and not fmt.Errorf. Thanks for the clarification!

**Review by swetharepakula — COMMENTED** [review]
/lgtm
/approve

**Pull request merged**